Enhance and add Python bindings for detected_object_set_output

neal-siekierski opened the pull request:
This PR does two main things:
- Add a "`complete`" method to `detected_object_set_output` and have `detected_object_output_process` call it at the end of the input stream
- Add Python bindings to allow subclassing `detected_object_set_output`

It also fixes an issue with `detected_object_set_ouput`'s destructor.

The motivation for these changes is to allow writing an arrow for JSON (specifically, MS-COCO) output. I have a working prototype written using these changes.

As it stands on its own, I have the following questions/concerns about the changes:
- ~Is this the right way to detect the end of the input stream? (I essentially copied the changes from `compute_track_descriptors_process`.)~
  - *(Edit 2020-02-14):* Changed to use `_finalize`
- ~The wrapping is not complete; rather I only wrapped what I needed and `static_type_name`.~
  - *(Edit 2020-02-14):* `open` and `close` were added on Jan. 28
- `stream`, which is how one writes output, was awkward to wrap because it returns an `ostream`. I went for returning a file-like object. Perhaps `detected_object_set_output` could have a more wrappable interface?
  - *(Edit 2020-02-14):* `stream` is no longer wrapped (overriding `open` and `close` is preferred), though the question remains

~Also, it looks like @as6520 has made overlapping changes in [`dev/add-activity-detector`](https://github.com/Kitware/kwiver/compare/master...dev/add-activity-detector), namely b51fe5c4a and 30b0ec88c. Regarding the relation between these changes and those:~
- ~The other changes wrapped `open` and `close`, while I did not.~
  - ~These didn't seem useful to wrap, at least from the perspective of subclassing `detected_object_set_output`, since subclassing code isn't expected to call them.~
- ~The other changes made `open` and `close` virtual.~

*(Edit 2020-02-14):* Per above, things have largely been reconciled, though #925 still has overlapping changes.

I think it would be good to settle on a selection of features we want to add to `detected_object_set_output` and have in its bindings and then get that merged in. This is a smaller set of changes than `dev/add-activity-detector`, so perhaps that could be done using this PR. @as6520, how does that work for you? (Similarly, I plan to look at creating a MS-COCO reader as well, which I expect to involve a similar discussion.)

as6520 commented:
I have not tested the binding yet. I sort of started binding all the abstract algorithms. Its not complete but my work is in dev/bind-algorithms. I merged the algorithms that I had created bindings for in dev/add-activity-detector.

as6520 commented:
Since  I am not a consumer of the algorithm, I am not attached to the changes.

as6520 commented on vital/algo/detected_object_set_output.h:
```diff
@@ -110,6 +110,12 @@ class VITAL_ALGO_EXPORT detected_object_set_output
+  /**
+   * This method writes any necessary final data to the currently open file.
+   */
+  virtual void complete() {}
```
This might cause problems with two implementations of detected_object_set_output in arrows/core since neither arrows are overriding it.

neal-siekierski replied:
They shouldn't need to override it as they don't need to do anything special when the input ends. Or am I missing something?

as6520 replied:
No you are correct.

as6520 commented:
@neal-siekierski If we make open and close virtual, provide trampolines for them it would make it easier for python arrows to sidestep streams since you can override those function in python. In either case we should bind them. The simplest use case for the bindings is using arrows in python without sprokit. It would look like 
```python
from kwiver.vital.modules import load_known_modules
from kwiver.vital.algo import DetectedObjectSetOutput
from kwiver vital.types import DetectedObjectSet

load_known_modules()
# Create the algorithm
doso = DetectedObjectSetOutput.create('csv')
doso.open('test_file.csv')
doso.write_set('DetectedObjectSet())
doso.close()
```
Thoughts?

neal-siekierski commented on sprokit/processes/core/detected_object_output_process.cxx:
```diff
@@ -157,6 +160,18 @@ ::_init()
 void detected_object_output_process
 ::_step()
 {
+  auto datum = peek_at_datum_using_trait( detected_object_set );
```
Compare [`compute_track_descriptors_process`](https://github.com/Kitware/kwiver/blob/00eb2cec299fc193f87081c65ff9400c5bf9d033/sprokit/processes/core/compute_track_descriptors_process.cxx#L149), which I modeled this code after.

For `_step` to be called in the block you linked, `d->is_complete` needs to be false and `d->check_required_input()` needs to return null. The former appears to be set only by calling `mark_process_as_complete`, and the latter directly returns null if [`check_input_level == check_none`](https://github.com/Kitware/kwiver/blob/46afabc100ba55e24391fadb710ed2cf05d61989/sprokit/src/sprokit/pipeline/process.cxx#L2054), where `check_input_level` is set by `set_data_checking_level`.

The main concern I have with the code I wrote (and therefore also with the existing `compute_track_descriptors_process`) is that it doesn't handle the other datum types (`empty`, `flush`, `error`, and `invalid`) at all, apparently effectively assuming that they'll never occur.

as6520 replied:
Aha, didn't realize that `check_input_level` did that

as6520 replied:
Aha, didn't know that. See below for the my ideas about handling other datum types.

neal-siekierski commented:
> @neal-siekierski If we make open and close virtual, provide trampolines for them it would make it easier for python arrows to sidestep streams since you can override those function in python.

That's a fair point. Certainly for `detected_object_set_input` sidestepping would be better than trying to recreate all the methods a readable file object normally has, and symmetry between the two would be good. I find the need to sidestep somewhat unfortunate, though. The current (would-become-)default implementation of `open` and `close` would perhaps be better in a subclass of `detected_object_set_output` that the C++ implementations could in turn extend from, though I'm not sure that's worth doing, particularly here.

> In either case we should bind them. The simplest use case for the bindings is using arrows in python without sprokit.

Thanks for the use case. That makes sense and I do agree we should bind the other methods.

> Thoughts?

See above. I'll add wrapping for the other methods and make `open` and `close` virtual, but I'll remove the wrapping for `stream`. I'll also update my prototype JSON writer implementation to use the sidestepping approach and then I'll test.

as6520 commented:
@neal-siekierski I agree that wrapping streams would be the ideal solution. There is some support for it as documented [here](https://pybind11.readthedocs.io/en/stable/advanced/pycpp/utilities.html#capturing-standard-output-from-ostream). But I don't think it is our intended use-case.

Going back to the original concerns about handling different datum types, I imagine there are two approaches to solve this issue
1.  If it is possible to move the logic that you have in `complete` to `close` use it. As close would be called in the destructor of `detected_object_set_output` your logic would be executed without having to worry about datums. Note: if we do this then complete would become a redundant method.

2. Check for datum types and if it is not `complete` or `data`, absorb the input and forward the datum to the edges as sprokit is doing [here](https://github.com/Kitware/kwiver/blob/master/sprokit/src/sprokit/pipeline/process.cxx#L368). I would also recommend updating the input level to valid in this case.

neal-siekierski commented:
I've added the wrapping for `open` and `close` and dropped it for `stream`, and I've successfully adapted my prototype writer to use the new API. Unfortunately I don't think overridden `close` can be guaranteed to be called on destruction. (Though perhaps it's worth calling it from the destructor of `detected_object_output_process`.)

On a process note, I'll rewrite these commits once the branch is more settled.

> Going back to the original concerns about handling different datum types, I imagine there are two approaches to solve this issue
> 
> 1. If it is possible to move the logic that you have in `complete` to `close` use it. As close would be called in the destructor of `detected_object_set_output` your logic would be executed without having to worry about datums. Note: if we do this then complete would become a redundant method.

I don't think this is the right approach. It probably wouldn't work because virtual methods don't dispatch like normal during destruction (see e.g. [here](https://www.artima.com/cppsource/nevercall.html) for a reference). It's also bad practice to run failable code in a destructor (another reference [here](https://isocpp.org/wiki/faq/exceptions#dtors-shouldnt-throw)), which arbitrary Python code should probably be considered to be.

> 2. Check for datum types and if it is not `complete` or `data`, absorb the input and forward the datum to the edges as sprokit is doing [here](https://github.com/Kitware/kwiver/blob/master/sprokit/src/sprokit/pipeline/process.cxx#L368). I would also recommend updating the input level to valid in this case.

It looks like the available [checking levels](https://github.com/Kitware/kwiver/blob/00eb2cec299fc193f87081c65ff9400c5bf9d033/sprokit/src/sprokit/pipeline/process.h#L299) are "none", "sync", and "valid". "Valid" only exposes `data`-type datums. A cursory look suggests the "sync" level might be an improvement over using "none", but it wasn't entirely clear. Also, as a minor hurdle to correctly reproducing some of `process::step`'s logic, `process::priv`'s methods are only declared in the `.cxx` file. I'd appreciate an official recommendation from @linus-sherrill or the like.

as6520 commented:
@neal-siekierski #921 added `_finalize` method to the sprokit process. This would be called after the all  `_step` functions have been called.  `complete` can be called in the overridden version of `detected_object_output_process`

neal-siekierski commented:
I've adjusted this to use the new overridable `_finalize`. I also changed the relevant declarations in `detected_object_output_process.h` to use `override` instead of `virtual` where relevant.

as6520 commented:
LGTM

as6520 commented:
@neal-siekierski The changes in #925 are originating from dev/bind-algorithms. Those branches would have be rebased on master to make the python tests work again. We can rebase after #919 lands and accept all changes made here as is. Thus these bindings would be the beings for detected_object_set_output

linus-sherrill commented:
Jenkins, test this please
